### PR TITLE
webadmin: improve audit log for user profile

### PR DESCRIPTION
**Pull request by rszwajko**
Before, a generic audit log message was used for all operations on
user profile virtual entity. The message included only the name of the
user that performed the operation.

Changes:
1. include target user ID - use case: admin creating/deleting
   properties for a regular user
2. include property name
3. use different messages for create/remove/replace operations

Notes about implementation:
1. prefix `Account Settings:` was added for Notification Drawer. In the audit log we have more information i.e. message code. In the UI this context is missing so message needs to be explicit.
2. ~~the current DB layer does not provide the name of the user linked with the property - we have only user ID. Retrieving the user name can be done under different PR.~~

### CODE REVIEW

**Comment by rszwajko** [comment]
Example logs:
```
2022-03-09 13:24:51,913+01 INFO  [org.ovirt.engine.core.dal.dbbroker.auditloghandling.AuditLogDirector] (default task-3) [c56549de-1bea-44a2-8bd5-5b731bda98da] EVENT_ID: USER_PROFILE_REMOVE_PROPERTY(878), Account Settings: admin@internal-authz removed property 'vmPortal.refreshInterval' for user 1cdf57d4-1179-11ec-8563-52540012eda2.
2022-03-09 13:24:52,162+01 INFO  [org.ovirt.engine.core.dal.dbbroker.auditloghandling.AuditLogDirector] (default task-3) [4e461f00-f3ba-4f92-8d8b-032cad0d23eb] EVENT_ID: USER_PROFILE_CREATE_PROPERTY(876), Account Settings: admin@internal-authz created property 'vmPortal.refreshInterval' for user 1cdf57d4-1179-11ec-8563-52540012eda2.
2022-03-09 13:26:22,740+01 INFO  [org.ovirt.engine.core.dal.dbbroker.auditloghandling.AuditLogDirector] (default task-3) [5467f362-d961-4334-83db-f600ceacbbce] EVENT_ID: USER_PROFILE_REPLACE_PROPERTY(880), Account Settings: admin@internal-authz replaced property 'SSH_PUBLIC_KEY' for user 1cdf57d4-1179-11ec-8563-52540012eda2.

```

**Review by sgratch — CHANGES_REQUESTED** [review]
The whole idea of adding those audit log messages is for tracking the management of properties settings. But since our behavior with those properties is not consistent comparing admin/vm portals and also comparing between the different properties, it's a bit hard to understand which operation was actually done by the user on a higher level (update a property, reset a property or all properties, created one?).

For example:
1. When updating an existing user property via the VM portal, it always break down to: `removed` + `created` the property. While on admin portal it always break down into `replaced` the property only. This is confusing and bottom line doesn't really reflects a consistent operation.

2. When resetting all settings - not all properties are actually removed. Part of them are set back to a default value. so the '`Rest Settings`' operation is broke down into: removing few of properties but updating few others. And then when a user set them again - part of them are created and part are updated.  So there is no way to know that the user was actually resetting back to defaults. 

Bottom line, logging from the `commands` level is very confusing and inconsistent. i wonder if there is a way to generate those logs on a higher level for the vm portal and for reset settings at least.

File: backend/manager/modules/dal/src/main/resources/bundles/AuditLogMessages.properties [hunk 1]
```diff
@@ -482,6 +482,12 @@ USER_REMOVE_ROLE=Role ${RoleName} removed from ${UserName}
 USER_REMOVE_ROLE_FAILED=Failed to remove role ${RoleName} (User ${UserName})
 USER_UPDATE_PROFILE=Updated profile for ${UserName}
 USER_UPDATE_PROFILE_FAILED=Failed to update profile for ${UserName}
+USER_PROFILE_CREATE_PROPERTY=Account Settings: ${UserName} created property '${ProfileProperty}' for user ${ProfileUser}.
+USER_PROFILE_CREATE_PROPERTY_FAILED=Account Settings: ${UserName} failed to create property '${ProfileProperty}' for user ${ProfileUser}.
+USER_PROFILE_REMOVE_PROPERTY=Account Settings: ${UserName} removed property '${ProfileProperty}' for user ${ProfileUser}.
+USER_PROFILE_REMOVE_PROPERTY_FAILED=Account Settings: ${UserName} failed to remove property '${ProfileProperty}' for user ${ProfileUser}.
+USER_PROFILE_REPLACE_PROPERTY=Account Settings: ${UserName} replaced property '${ProfileProperty}' for user ${ProfileUser}.
+USER_PROFILE_REPLACE_PROPERTY_FAILED=Account Settings: ${UserName} failed to replace property '${ProfileProperty}' for user ${ProfileUser}.
```
[the remark this answers is not in the document]
Comment (rszwajko) [reply backend/manager/modules/dal/src/main/resources/bundles/AuditLogMessages.properties:490]: > When you write replace, it's not clear if replaced by what.
>

Replaced by another property with the same name. 

> Maybe using the term 'update' is better here since what's really interesting is that the value was updated.
>

We cannot use `update` as this suggest mutating the state of the same entity - in this case a new immutable entity is created.

**Comment by rszwajko** [comment]
> it's a bit hard to understand which operation was actually done by the user on a higher level (update a property, reset a property or all properties, created one?).
> 

Audit log was not key priority. The properties do not impact engine operation or resource allocation and are limited to improving user experience in the UI. If we have a use case where admin needs to understand this kind of changes then we can improve audit capabilities.


> For example:
> 
> 1. When updating an existing user property via the VM portal, it always break down to: `removed` + `created` the property. While on admin portal it always break down into `replaced` the property only. This is confusing and bottom line doesn't really reflects a consistent operation.
>

Audit log is a low level log that reflects the changes on the backend. The same flow in the UI is implemented differently for REST API and for GWT. The new REST API cannot use "replace" as "update" so it uses "delete" + "add" instead. The GWT version can access the command layer directly and is able to take advantage of "replace" command.
 
> 2. When resetting all settings - not all properties are actually removed. Part of them are set back to a default value. so the '`Rest Settings`' operation is broke down into: removing few of properties but updating few others.
>
The exact flow of reset settings operation is an implementation detail. It might be implemented by removal of the property or by replacing with property that has "default" value. Having said that, all currently supported properties use "remove" approach. If you have noticed different approach then please let me know.

> And then when a user set them again - part of them are created and part are updated.  So there is no way to know that the user was actually resetting back to defaults.
> 

The reset is limited to (chosen) properties displayed in the Account Setting dialog - other settings are no affected by design.
The changes are listed in the confirmation dialog and later on reflected in the dialog itself. This works the same as "reset settings" in VM Portal. Please provide more details on how the flow should look like in your opinion.

> Bottom line, logging from the `commands` level is very confusing and inconsistent. i wonder if there is a way to generate those logs on a higher level for the vm portal and for reset settings at least.
>

As mentioned above, we are abusing the audit log here: it simply reflects the true flow executed on the backend(for audit purpose). This doesn't match the UI flow but having 1:1 mapping between UI and backend is not one of the project goals.

File: backend/manager/modules/dal/src/main/resources/bundles/AuditLogMessages.properties (hunk 1 above)
[the remark this answers is not in the document]
Comment (sgratch) [reply backend/manager/modules/dal/src/main/resources/bundles/AuditLogMessages.properties:490]: I know, but users won't surely understand WDYM by replace if they won't dig into DB queries, so maybe rephrasing to something  like:
 "userxx regenerated (for updating) property xxx for user...."

[the remark this answers is not in the document]
Comment (rszwajko) [reply backend/manager/modules/dal/src/main/resources/bundles/AuditLogMessages.properties:490]: `regenerated (for updating)` suggest mutation which is incorrect since a new (immutable) entity is created.

File: backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/AddUserProfilePropertyCommand.java [hunk 2]
```diff
@@ -28,6 +31,8 @@ public AddUserProfilePropertyCommand(T parameters, CommandContext cmdContext) {
     @Override
     protected boolean validate() {
         UserProfileProperty newProp = getParameters().getUserProfileProperty();
+        addCustomValue(PROFILE_USER, newProp.getUserId().toString());
```
[the remark this answers is not in the document]
Comment (sgratch) [reply backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/AddUserProfilePropertyCommand.java:34]: Since `user ID` is not displayed as part of webadmin general info, it's much less convenient for admins to figure out which user is referred by this message. Note that `user ID` is not used on most of audit log messages while other entities does appear - since this info is exposed on webadmin.

I prefer to avoid that part of adding the target user till we find a way to solve it.

Or, if you insists :), just put that within parenthesis as an extra info for users that do want to query the DB for that.
Maybe, you can also add `property id` within those parenthesis as well in that case.

[the remark this answers is not in the document]
Comment (sgratch) [reply backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/AddUserProfilePropertyCommand.java:34]: > Bottom line, logging from the `commands` level is very confusing and inconsistent. i wonder if there is a way to generate those logs on a higher level for the vm portal and for reset settings at least.

@rszwajko what about moving to a upper level for logging a real user actions like "updating a property" or "resetting a property"? This has much more value than just logging each command...

**Comment by rszwajko** [comment]
@sgratch 
Please check the newest version - the log message now contains target user name if needed. The implementation is based on extra query to fetch target user.

**Review by sgratch — COMMENTED** [review]
@rszwajko 

> regenerated (for updating) suggest mutation which is incorrect since a new (immutable) entity is created.

Correct, but the admin user is not necessarily a DBA so he might not understand the meaning as we do. Replacing means changing the property with what? Messages should be accurate but low level details are sometimes confusing and not enough. Maybe just add a description of replacing with what, e.g.
${UserName} replaced property '${profile_property} with the same property name and an updated value.

> Please check the newest version - the log message now contains target user name if needed. The implementation is based on extra query to fetch target user.

+1, used by other command as well and by audit log too so I'm ok with that.

> Audit log was not key priority. The properties do not impact engine operation or resource allocation and are limited to improving user experience in the UI. If we have a use case where admin needs to understand this kind of changes then we can improve audit capabilities.

The point is that based on logs it's impossible for admins and for trouble shooting to understand what the user configured with user settings up till now and I think it's needed. We could implement "reset" as part of REST as well if needed instead of asking the user to reset one by one via the apis.  That's not an excuse for the problem of tracking user actions. Most of user actions are logged (at least on DEBUG log level) so that it will be easy to understand what the user tried to do.

But since it will require extra effort to enable that on a upper level now and since webadmin and web ui are different then let's leave it as is for now, even though not ideal.

> The exact flow of reset settings operation is an implementation detail. It might be implemented by removal of the property or by replacing with property that has "default" value. Having said that, all currently supported properties use "remove" approach. If you have noticed different approach then please let me know.

This is correct for webadmin properties and not correct for all vmportal ones. All vmportal properties are not just removed when calling reset settings, they are removed and then re-created with the default value. While for webadmin all user's properties are removed except `SSH_PUBLIC_KEY` and `webadmin` (grids setting) . Can we at least keep consistency on that level?

File: backend/manager/modules/common/src/main/java/org/ovirt/engine/core/common/AuditLogType.java [hunk 3]
```diff
@@ -1076,12 +1076,25 @@
     USER_ADD_SYSTEM_PERMISSION_FAILED(867, AuditLogSeverity.ERROR),
     USER_REMOVE_SYSTEM_PERMISSION(868),
     USER_REMOVE_SYSTEM_PERMISSION_FAILED(869, AuditLogSeverity.ERROR),
+    // left for accessing legacy entries
```
[the remark this answers is not in the document]
Comment (rszwajko) [reply backend/manager/modules/common/src/main/java/org/ovirt/engine/core/common/AuditLogType.java:1080]: Actually the comment applies to the whole block. Added new lines to separate them from the rest of the code

**Comment by rszwajko** [comment]
> Maybe just add a description of replacing with what, e.g.
> ${UserName} replaced property '${profile_property} with the same property name and an updated value.
>

New message uses `instance` to differentiate between old and new property. Note that only the object identity is guaranteed to change - the value may stay the same.

> But since it will require extra effort to enable that on a upper level now and since webadmin and web ui are different then let's leave it as is for now, even though not ideal.
>

+1



>> The exact flow of reset settings operation is an implementation detail. It might be implemented by removal of the property or by replacing with property that has "default" value. Having said that, all currently supported properties use "remove" approach. If you have noticed different approach then please let me know.
>>
> This is correct for webadmin properties and not correct for all vmportal ones. All vmportal properties are not just removed when calling reset settings, they are removed and then re-created with the default value. While for webadmin all user's properties are removed except SSH_PUBLIC_KEY and webadmin (grids setting) . Can we at least keep consistency on that level?
>

This is definitely out of scope of this PR. Please create another issue and let's discuss it there.

File: backend/manager/modules/common/src/main/java/org/ovirt/engine/core/common/AuditLogType.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (rszwajko) [reply backend/manager/modules/common/src/main/java/org/ovirt/engine/core/common/AuditLogType.java:1080]: `@Deprecated` looks as an even better alternative.

**Comment by sgratch** [comment]
> New message uses `instance` to differentiate between old and new property. Note that only the object identity is guaranteed to change - the value may stay the same.

+1, much better

> This is definitely out of scope of this PR. Please create another issue and let's discuss it there.

This is definitely out of scope, it is just reflected by this fix more easily.
I opened a new issue: https://github.com/oVirt/ovirt-web-ui/issues/1576

**Review by sgratch — APPROVED** [review]
LGTM
